### ISSUE TITLE
SDK install gives HTTP 407 (Proxy authentication required)

### ISSUE THREAD

**Issue by tacker66** [git-for-windows/git#309]
Same problems e.g. with conda package manager or NetBeans updater.
The mechanism git itself uses (http.proxy) for fetching through the proxy
works.

**Comment by dscho** [comment]
Thank you for opening a new ticket (this will make it much easier to find for others with similar issues).

As to 407, it seems that this means the proxy authentication failed: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#407

So... can you tell us a bit more context, in particular about your proxy setup?

**Comment by tacker66** [comment]
I have no further information about this proxy but as mentioned above Git 2.5.0 (Windows and Cygwin) somehow manage to download files through the proxy anonymously.

**Comment by dscho** [comment]
Hmm. So maybe Pacman has problems to install with the proxy? But you specified the proxy when installing the SDK, yes?

Maybe it would be easier to help you if you could paste the output of the SDK when it fails. (If there is too personal info, please edit instead of removing)

**Comment by tacker66** [comment]
This is what I get when running git-sdk-installer-1.0.0-64.7z.exe: 

	:: Synchronizing package databases...
	error: failed retrieving file 'git-for-windows.db' from dl.bintray.com : Connection timed out after 10001 milliseconds
	error: failed to update git-for-windows (download library error)
	error: failed retrieving file 'mingw32.db' from repo.msys2.org : Connection timed out after 10000 milliseconds
	error: failed retrieving file 'mingw32.db' from downloads.sourceforge.net : Connection timed out after 10000 milliseconds
	error: failed retrieving file 'mingw32.db' from www2.futureware.at : Connection timed out after 10000 milliseconds
	error: failed to update mingw32 (download library error)
	error: failed retrieving file 'mingw64.db' from repo.msys2.org : Connection timed out after 10000 milliseconds
	error: failed retrieving file 'mingw64.db' from downloads.sourceforge.net : Connection timed out after 10000 milliseconds
	error: failed retrieving file 'mingw64.db' from www2.futureware.at : Connection timed out after 10000 milliseconds
	error: failed to update mingw64 (download library error)
	error: failed retrieving file 'msys.db' from repo.msys2.org : Connection timed out after 10000 milliseconds
	error: failed retrieving file 'msys.db' from downloads.sourceforge.net : Connection timed out after 10000 milliseconds
	error: failed retrieving file 'msys.db' from www2.futureware.at : Connection timed out after 10000 milliseconds
	error: failed to update msys (download library error)
	error: failed to synchronize any databases
	error: failed to init transaction (download library error)

	There was a problem accessing the MSys2 repositories
	If your setup requires an HTTP proxy to access the web,
	please specify it here, otherwise leave it empty.

	HTTP proxy: <proxyname:proxyport>

	:: Synchronizing package databases...
	error: failed retrieving file 'git-for-windows.db' from dl.bintray.com : Received HTTP code 407 from proxy after CONNECT
	error: failed to update git-for-windows (download library error)
	error: failed retrieving file 'mingw32.db' from repo.msys2.org : The requested URL returned error: 407
	error: failed retrieving file 'mingw32.db' from downloads.sourceforge.net : The requested URL returned error: 407
	error: failed retrieving file 'mingw32.db' from www2.futureware.at : Received HTTP code 407 from proxy after CONNECT
	error: failed to update mingw32 (download library error)
	error: failed retrieving file 'mingw64.db' from repo.msys2.org : The requested URL returned error: 407
	error: failed retrieving file 'mingw64.db' from downloads.sourceforge.net : The requested URL returned error: 407
	error: failed retrieving file 'mingw64.db' from www2.futureware.at : Received HTTP code 407 from proxy after CONNECT
	error: failed to update mingw64 (download library error)
	error: failed retrieving file 'msys.db' from repo.msys2.org : The requested URL returned error: 407
	error: failed retrieving file 'msys.db' from downloads.sourceforge.net : The requested URL returned error: 407
	error: failed retrieving file 'msys.db' from www2.futureware.at : Received HTTP code 407 from proxy after CONNECT
	error: failed to update msys (download library error)
	error: failed to synchronize any databases
	error: failed to init transaction (download library error)

**Comment by dscho** [comment]
Could you try with an empty user name and password in the Proxy specification? I.e. if your proxy is `http://proxy.internal:8080/`, try specifying `http://:@proxy.internal:8080/`, i.e. prefix the host name with a `:@`.

**Comment by tacker66** [comment]
Same result ...

**Comment by dscho** [comment]
What kind of proxy and proxy authentication do you use?

**Comment by tacker66** [comment]
Seems to be an NTLM proxy.

**Comment by dscho** [comment]
@tacker66 you would not happen to know *which* NTLM proxy (I am trying to reproduce here...).

**Comment by tacker66** [comment]
Sorry, this is not really my area of expertise ... what do you mean by "which" proxy?
(It's a corporate proxy which seems to use NTLM authorization and is not accessible
from the outside.)

**Comment by dscho** [comment]
Right. I have troubles replicating this scenario here... :-(

**Comment by dscho** [comment]
BTW which WIndows version is your client (just so I can try my best to make as similar a setup to yours here as possible)?

**Comment by madbonez** [comment]
Hi guys! I had same issue on versions (Only they tried) 2.5.0 and 2.4.6.5. Detected request by Wire Shark:

![image](https://cloud.githubusercontent.com/assets/7160484/9469178/4fc894b4-4b4e-11e5-829f-aa7f8994e6df.png)

Seems like Git trying use ntlm by default, but i am using Basic auth in my corp.
When i rollback on 2.3.5.8  everything is working. And Git attaching proper headers on request:

![image](https://cloud.githubusercontent.com/assets/7160484/9469284/e2fd9644-4b4e-11e5-9377-ab8003c38c00.png)

**Comment by dscho** [comment]
@tacker66 Actually, there is one thing I would like you to try first: could you download (manually) https://github.com/git-for-windows/build-extra/releases/download/tmp-sdk-installer-extra/curl-extra.zip and unpack it into the SDK installer directory (there should still be `usr\bin\pacman.exe`; make sure that the files are unpacked with full path, i.e. into `usr\bin\`), then edit the file `usr\etc\pacman.conf` and remove the `#` in front of the `XferCommand` line mentioning `/usr/bin/curl`, and then run (from `cmd.exe`, in that `usr\bin\` directory, with `http_config` and `https_config` specifying your NTLM proxy):

```sh
pacman -Sy
```

and paste the output here?

**Comment by dscho** [comment]
@madbonez I fear you mistook this issue; it is not about Git itself, but about the SDK installer (and the reporter [actually mentioned that 2.5.0 works for them](https://github.com/git-for-windows/git/issues/309#issuecomment-134237119)...

**Comment by tacker66** [comment]
@dscho: where/how do I specify http_config/https_config? In the environment as in 
"set https_config=https://proxy-emea-fth.inac.siemens.net:84" ?

**Comment by tacker66** [comment]
@dscho: how do I specify http_config/https_config? In the environment?

**Comment by dscho** [comment]
In the environment, as it is done in the SDK installer: https://github.com/git-for-windows/build-extra/blob/350624b44b3224605d0da0a7f8a8ceb1b1d796d4/sdk-installer/setup-git-sdk.bat#L34-L38

**Comment by tacker66** [comment]
Here the result (theres is no usr/etc/pacman.conf but a etc/pacman.conf):

	set http_config=http://<name>:<port>

	set https_config=https://<name>:<port>

	pacman -Sy

	:: Synchronizing package databases...
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	  0     0    0     0    0     0      0      0 --:--:--  0:00:16 --:--:--     0

**Comment by dscho** [comment]
Is your proxy a HTTPS proxy or an HTTP proxy? You should use that protocol, not different ones for the two environment variables. The `http_` and `https_` prefix only state which URLs should be retrieved via the proxy, not *how* you want to talk to the proxy.

Also, could you specify `http://:@<name>:<port>/`, as I suggested earlier? The `:@` -- an empty user name and an empty password -- should be interpreted by `curl.exe` to use NTLM authentication.

**Comment by tacker66** [comment]
Same result with

	  set http_config=https://:@<name>:<port>
	  set https_config=https://:@<name>:<port>

**Comment by dscho** [comment]
Bummer.

So if you start `Git Bash` (not in the Git SDK), does `curl https://github.com/` work correctly, *without* setting `https_config`?

**Comment by tacker66** [comment]
With "Git Bash here" same result as above (an also the same when setting https_config).
(But git fetch from e.g. https://github.com/git/git.git works)

**Comment by dscho** [comment]
Oooh, I think we're getting somewhere. Git configures a couple of flags to be able to use NTLM authentication IIRC. So that sent me on a StackOverflow search, and it looks as if [this post](http://stackoverflow.com/a/17442412) has the answer:

```sh
curl --proxy-ntlm --proxy %PROXY_URL%:%PORT% %SITE_URL%
```

Does that work in Git Bash? If so, does it work if you set `https_proxy` and pass only the `--proxy-ntlm` option (i.e. *not* the `--proxy` option)? If so, could you add the `--proxy-ntlm` option to the `XferCommand` in your `etc\pacman.conf` and test whether `pacman -Sy` works now?

**Comment by tacker66** [comment]
curl --proxy-ntlm --proxy https://<proxyname>:<proxyport> https://github.com
works :-) But setting the proxy via environment doesn't :-(

**Comment by dscho** [comment]
Okay, let's go one step further... Can you edit the `XferCommand` so that it has the `--proxy` option, too?

**Comment by tacker66** [comment]
I didn't notice, that you silently switched from "https_config" to "https_proxy" ;-)
With "https_proxy" it works. "pacman -Sy" gives

	:: Synchronizing package databases...
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100  2424  100  2424    0     0   8728      0 --:--:-- --:--:-- --:--:--  9431
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (22) The requested URL returned error: 404 Not Found
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100  196k  100  196k    0     0  2615k      0 --:--:-- --:--:-- --:--:-- 2894k
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100    96  100    96    0     0   5148      0 --:--:-- --:--:-- --:--:--  8000
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100  197k  100  197k    0     0  2857k      0 --:--:-- --:--:-- --:--:-- 3180k
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100    96  100    96    0     0   4075      0 --:--:-- --:--:-- --:--:--  6857
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100  124k  100  124k    0     0  2210k      0 --:--:-- --:--:-- --:--:-- 2489k
	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
									 Dload  Upload   Total   Spent    Left  Speed
	100    96  100    96    0     0   5278      0 --:--:-- --:--:-- --:--:--  8727

**Comment by dscho** [comment]
Aha! So it works ;-)

Yeah, the `s` in `https_proxy` is required because both Git for Windows' and MSys2's Pacman repositories are accessed via HTTPS.

The only unclear thing before we can go and implement a fix (which I will ask you to test!) is: does `--ntlm-proxy` prevent non-NTLM proxies from working...

**Comment by dscho** [comment]
Ooooh! I just found the `--anyauth` option... Win!

**Comment by dscho** [comment]
So... @tacker66 it is testing time! Please download the (temporary) installer from https://github.com/git-for-windows/build-extra/releases/download/tmp-sdk-installer-extra/git-sdk-installer-1.0.1-64.7z.exe -- I would like you to install it and report back whether it

- still times out the first time
- asks for the proxy
- works after you type in the proxy

Thanks!

**Comment by tacker66** [comment]
After timing out the first time it asks for the proxy and then fails with

			1 Datei(en) kopiert.
	:: Synchronizing package databases...
	error: failed to update git-for-windows (error invoking external downloader)
	error: failed to update mingw32 (error invoking external downloader)
	error: failed to update mingw64 (error invoking external downloader)
	error: failed to update msys (error invoking external downloader)
	error: failed to synchronize any databases
	error: failed to init transaction (error invoking external downloader)

	There was a problem accessing the MSys2 repositories
	If your setup requires an HTTP proxy to access the web,
	please specify it here, otherwise leave it empty.

	HTTP proxy:

**Comment by dscho** [comment]
Hmm. Could you open a `cmd`, `cd` to that directory and then call

```cmd
usr\bin\pacman -Syv --debug --verbose
```

and paste the output here?

**Comment by tacker66** [comment]
I tried again with a clean install, and now it worked :-)

**Comment by dscho** [comment]
Yeah! Thank you for reporting back!

**Comment by dscho** [comment]
And new SDK installers have been uploaded, too: https://github.com/git-for-windows/build-extra/releases/tag/git-sdk-1.0.1
